### DISCUSSION TITLE
Execute code after setup node events and before the spec starts execution

### DISCUSSION THREAD

**Discussion by roshnidevadiga** [cypress-io/cypress#27811] [Questions and Help]
I want to execute a setup code just before my spec execution starts. If I do it in setup node events before run hook, I cannot execute any cy commands since it does not execute in the context of browser. How do I achieve this? Is there any other hook available?

**Comment by portyu9** [comment]
Hi @roshnidevadiga 

If the setup requires `cy.*` commands, it should not go in `setupNodeEvents()`.

`setupNodeEvents`, `before:run`, and `before:spec` execute in the **Node process**, so Cypress commands such as `cy.request()`, `cy.visit()`, etc. are unavailable there.

For browser-side setup, put a Mocha hook in your Cypress support file.

Once per spec:

```js
// cypress/support/e2e.js

before(() => {
  cy.request('/some/setup')
  // other Cypress commands
})
```

Before every test:

```js
// cypress/support/e2e.js

beforeEach(() => {
  cy.request('/some/setup')
})
```

The support file is loaded for each spec and root-level hooks defined there apply to the tests in that spec.

If what you need to execute is genuinely Node-side code and must happen immediately before Cypress launches/runs a spec, then `before:spec` is the correct event:

```js
setupNodeEvents(on, config) {
  on('before:spec', async (spec) => {
    // Node APIs are available here
    // cy.* commands are not
  })
}
```

One important distinction: a `before()` hook runs before the tests execute, but it does not run before JavaScript in the spec file is evaluated. There is no point before spec-file evaluation where you can enqueue `cy.*` commands.

So use `before:spec` for Node work and support-file `before()` / `beforeEach()` for Cypress commands.

Please mark this answer as accepted if it helped, thank you.
